README.md
# gRPC PostgreSQL CRUD Sample

Step 1. Create Proto file, define messages and services

Step 2. Specify appropriate inject tag comments on top of fields for custom gorm tags

Step 3. Generate pg.go file

Step 4. Inject Gorm Tag to skip xxx fields protoc-go-inject-tag -input=./postgrescrud/postgrescrud.pb.go -XXX_skip=gorm

Step 5. Generate Reverse proxy, pb.gw.go file

Step 6. Success!


CHANGES.md
# Changelog

### [0.0.3](https://www.github.com/atreya2011/grpc-postgres-crud/compare/v0.0.2...v0.0.3) (2021-06-07)


### Bug Fixes

* action ([f1faab8](https://www.github.com/atreya2011/grpc-postgres-crud/commit/f1faab8418d57ed3a6808e91256132342ed24728))
* changelog file ([327c971](https://www.github.com/atreya2011/grpc-postgres-crud/commit/327c9715b89c5eb0df90e7b079fa1827e8accb72))
* changelog file name ([27da4b8](https://www.github.com/atreya2011/grpc-postgres-crud/commit/27da4b8fb6e318b7dfd1ab13cb893ec1c6cf04ea))
* changelog path ([aca831f](https://www.github.com/atreya2011/grpc-postgres-crud/commit/aca831fbb62d90cecca4a85fe70079ed00b072af))
* filename again ([e5fe4f4](https://www.github.com/atreya2011/grpc-postgres-crud/commit/e5fe4f48df40813b0327c0fee89ebad271dd2168))
* fix ([690f02e](https://www.github.com/atreya2011/grpc-postgres-crud/commit/690f02ed1044c2b2e958e62df44570a35580024b))


proxy.go
package main

import (
	"context"
	"flag"
	"log"
	"net/http"

	"github.com/atreya2011/grpc-postgres-crud/postgrescrud"

	"github.com/golang/glog"
	"github.com/grpc-ecosystem/grpc-gateway/runtime"
	"google.golang.org/grpc"
)

// end point means the ip and port on which grpc server is listening
var postgrescrudEndPoint = flag.String("endpoint", "localhost:7000", "endpoint of PostgresCrudService")

func run() error {
	// Step 1. define a context with cancel
	ctx := context.Background()
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	// Step 2. define the serve mux options here
	muxOpts := []runtime.ServeMuxOption{
		// Todo: check meaning of MIMEWildcard
		// OrigName set to false, json field names will be camelCase
		// EmitDefaults set to true, empty json fields will not be omitted
		runtime.WithMarshalerOption(runtime.MIMEWildcard, &runtime.JSONPb{OrigName: false, EmitDefaults: true}),
	}

	// Step 3. create a new server mux
	mux := runtime.NewServeMux(muxOpts...)
	// Step 4. create grpc dial options for connecting to grpc server
	opts := []grpc.DialOption{grpc.WithInsecure()}
	// Step 5. register the handler from endpoint (function in generated pb.gw.go file)
	// the following are passed into the function
	// context, new server mux, endpoint string, dial options
	err := postgrescrud.RegisterPostgresCrudHandlerFromEndpoint(ctx, mux, *postgrescrudEndPoint, opts)
	if err != nil {
		log.Fatalln(err)
	}

	log.Println("Listening...")
	// Step 6. Start API Server with ListenAndServe, pass in the serve mux
	// We launch a goroutine for handling requests
	done := make(chan struct{})
	go func() {
		http.ListenAndServe(":8082", mux)
		done <- struct{}{}
	}()

	<-done //wait finish groutine
	log.Println("Program exit")
	return nil
}

func main() {
	// flag.Parse parses the flags passed in the command line
	// if no flag is present, default values are used
	// if there is no flag.Parse, command line flags are not parsed
	flag.Parse()
	defer glog.Flush()

	if err := run(); err != nil {
		glog.Fatalln(err)
	}
}


server-1.go
package main

import (
	"context"
	"flag"
	"log"
	"net"

	"github.com/go-pg/pg/orm"

	"github.com/atreya2011/grpc-postgres-crud/postgrescrud"
	"github.com/go-pg/pg"
	"github.com/golang/protobuf/ptypes/empty"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
)

type service struct {
	DB *pg.DB
}

var grpcAddrFlag = flag.String("addr", ":7000", "Address host:port")

func main() {
	log.Printf("grpc server start on port %v", *grpcAddrFlag)
	// Step 1. listen for connections on tcp
	lis, err := net.Listen("tcp", *grpcAddrFlag)
	// Always handle errors
	if err != nil {
		log.Fatalf("frak")
	}
	// Step 2. Create a new grpc server instance
	srv := grpc.NewServer()
	// Step 3. Initialize the db and store it in service{}
	serv := initDB()
	// Step 4. Register the Service by passing the new server instance
	// and the server struct created above
	// The register function is present in pb.go
	postgrescrud.RegisterPostgresCrudServer(srv, serv)
	// Step 5. Serve the listener created above
	srv.Serve(lis)
}

// Create The following is the implementation of the Create service
// as defined in the proto file. It can be any implemention.
func (s *service) Create(ctx context.Context, req *postgrescrud.CreatePersonRequest) (*postgrescrud.CreatePersonResponse, error) {
	err := s.DB.Insert(req.Person)
	if err != nil {
		return nil, grpc.Errorf(codes.Internal, "Could not insert item into the database: %s", err)
	}
	return &postgrescrud.CreatePersonResponse{Id: 0}, nil
}

// Read The following is the implementation of the Read service
// as defined in the proto file. It can be any implemention.
func (s *service) Read(ctx context.Context, req *postgrescrud.ReadPersonRequest) (*postgrescrud.ReadPersonResponse, error) {
	// find person with id
	// initialize p with a point to person type
	p := new(postgrescrud.Person)
	// store the result in p using gorm.Where
	err := s.DB.Model(p).Where("id = ?", req.GetId()).First()
	if err != nil {
		log.Println(err)
	}
	// return response
	return &postgrescrud.ReadPersonResponse{Person: p}, nil
}

// List lists all the fullnames. Request is empty.
// Below is an example implementation
func (s *service) List(ctx context.Context, e *empty.Empty) (*postgrescrud.ListPeopleResponse, error) {
	// initialize slice of pointers to person
	var people []*postgrescrud.Person
	// pass this to gorm.Find() to fill it with results
	query := s.DB.Model(&people).Order("id ASC")
	err := query.Select()
	if err != nil {
		log.Println(err)
	}
	return &postgrescrud.ListPeopleResponse{Peoples: people}, nil
}

// Delete deletes record based on id
// below is an example implementation
// func (*service) Delete() {
// 	// Delete - delete product
// 	db.Delete(&product)
// }

func initDB() *service {
	// Connect to database
	db := pg.Connect(&pg.Options{
		User:     "atreya",
		Database: "atreya",
		Addr:     "localhost:5432",
	})
	// Drop table
	db.DropTable(&postgrescrud.Person{}, &orm.DropTableOptions{IfExists: true})
	// Create Table from person struct generated by gRPC
	db.CreateTable(&postgrescrud.Person{}, nil)

	// Create a new service instance and store the db
	s := &service{DB: db}

	// return the db
	return s
}


postgrescrud/postgrescrud.pb.gw.go
// Code generated by protoc-gen-grpc-gateway. DO NOT EDIT.
// source: postgrescrud/postgrescrud.proto

/*
Package postgrescrud is a reverse proxy.

It translates gRPC into RESTful JSON APIs.
*/
package postgrescrud

import (
	"context"
	"io"
	"net/http"

	"github.com/golang/protobuf/proto"
	"github.com/golang/protobuf/ptypes/empty"
	"github.com/grpc-ecosystem/grpc-gateway/runtime"
	"github.com/grpc-ecosystem/grpc-gateway/utilities"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/grpclog"
	"google.golang.org/grpc/status"
)

var _ codes.Code
var _ io.Reader
var _ status.Status
var _ = runtime.String
var _ = utilities.NewDoubleArray

func request_PostgresCrud_Create_0(ctx context.Context, marshaler runtime.Marshaler, client PostgresCrudClient, req *http.Request, pathParams map[string]string) (proto.Message, runtime.ServerMetadata, error) {
	var protoReq CreatePersonRequest
	var metadata runtime.ServerMetadata

	newReader, berr := utilities.IOReaderFactory(req.Body)
	if berr != nil {
		return nil, metadata, status.Errorf(codes.InvalidArgument, "%v", berr)
	}
	if err := marshaler.NewDecoder(newReader()).Decode(&protoReq.Person); err != nil && err != io.EOF {
		return nil, metadata, status.Errorf(codes.InvalidArgument, "%v", err)
	}

	msg, err := client.Create(ctx, &protoReq, grpc.Header(&metadata.HeaderMD), grpc.Trailer(&metadata.TrailerMD))
	return msg, metadata, err

}

func request_PostgresCrud_List_0(ctx context.Context, marshaler runtime.Marshaler, client PostgresCrudClient, req *http.Request, pathParams map[string]string) (proto.Message, runtime.ServerMetadata, error) {
	var protoReq empty.Empty
	var metadata runtime.ServerMetadata

	msg, err := client.List(ctx, &protoReq, grpc.Header(&metadata.HeaderMD), grpc.Trailer(&metadata.TrailerMD))
	return msg, metadata, err

}

func request_PostgresCrud_Read_0(ctx context.Context, marshaler runtime.Marshaler, client PostgresCrudClient, req *http.Request, pathParams map[string]string) (proto.Message, runtime.ServerMetadata, error) {
	var protoReq ReadPersonRequest
	var metadata runtime.ServerMetadata

	var (
		val string
		ok  bool
		err error
		_   = err
	)

	val, ok = pathParams["id"]
	if !ok {
		return nil, metadata, status.Errorf(codes.InvalidArgument, "missing parameter %s", "id")
	}

	protoReq.Id, err = runtime.Int32(val)

	if err != nil {
		return nil, metadata, status.Errorf(codes.InvalidArgument, "type mismatch, parameter: %s, error: %v", "id", err)
	}

	msg, err := client.Read(ctx, &protoReq, grpc.Header(&metadata.HeaderMD), grpc.Trailer(&metadata.TrailerMD))
	return msg, metadata, err

}

// RegisterPostgresCrudHandlerFromEndpoint is same as RegisterPostgresCrudHandler but
// automatically dials to "endpoint" and closes the connection when "ctx" gets done.
func RegisterPostgresCrudHandlerFromEndpoint(ctx context.Context, mux *runtime.ServeMux, endpoint string, opts []grpc.DialOption) (err error) {
	conn, err := grpc.Dial(endpoint, opts...)
	if err != nil {
		return err
	}
	defer func() {
		if err != nil {
			if cerr := conn.Close(); cerr != nil {
				grpclog.Infof("Failed to close conn to %s: %v", endpoint, cerr)
			}
			return
		}
		go func() {
			<-ctx.Done()
			if cerr := conn.Close(); cerr != nil {
				grpclog.Infof("Failed to close conn to %s: %v", endpoint, cerr)
			}
		}()
	}()

	return RegisterPostgresCrudHandler(ctx, mux, conn)
}

// RegisterPostgresCrudHandler registers the http handlers for service PostgresCrud to "mux".
// The handlers forward requests to the grpc endpoint over "conn".
func RegisterPostgresCrudHandler(ctx context.Context, mux *runtime.ServeMux, conn *grpc.ClientConn) error {
	return RegisterPostgresCrudHandlerClient(ctx, mux, NewPostgresCrudClient(conn))
}

// RegisterPostgresCrudHandlerClient registers the http handlers for service PostgresCrud
// to "mux". The handlers forward requests to the grpc endpoint over the given implementation of "PostgresCrudClient".
// Note: the gRPC framework executes interceptors within the gRPC handler. If the passed in "PostgresCrudClient"
// doesn't go through the normal gRPC flow (creating a gRPC client etc.) then it will be up to the passed in
// "PostgresCrudClient" to call the correct interceptors.
func RegisterPostgresCrudHandlerClient(ctx context.Context, mux *runtime.ServeMux, client PostgresCrudClient) error {

	mux.Handle("POST", pattern_PostgresCrud_Create_0, func(w http.ResponseWriter, req *http.Request, pathParams map[string]string) {
		ctx, cancel := context.WithCancel(req.Context())
		defer cancel()
		inboundMarshaler, outboundMarshaler := runtime.MarshalerForRequest(mux, req)
		rctx, err := runtime.AnnotateContext(ctx, mux, req)
		if err != nil {
			runtime.HTTPError(ctx, mux, outboundMarshaler, w, req, err)
			return
		}
		resp, md, err := request_PostgresCrud_Create_0(rctx, inboundMarshaler, client, req, pathParams)
		ctx = runtime.NewServerMetadataContext(ctx, md)
		if err != nil {
			runtime.HTTPError(ctx, mux, outboundMarshaler, w, req, err)
			return
		}

		forward_PostgresCrud_Create_0(ctx, mux, outboundMarshaler, w, req, resp, mux.GetForwardResponseOptions()...)

	})

	mux.Handle("GET", pattern_PostgresCrud_List_0, func(w http.ResponseWriter, req *http.Request, pathParams map[string]string) {
		ctx, cancel := context.WithCancel(req.Context())
		defer cancel()
		inboundMarshaler, outboundMarshaler := runtime.MarshalerForRequest(mux, req)
		rctx, err := runtime.AnnotateContext(ctx, mux, req)
		if err != nil {
			runtime.HTTPError(ctx, mux, outboundMarshaler, w, req, err)
			return
		}
		resp, md, err := request_PostgresCrud_List_0(rctx, inboundMarshaler, client, req, pathParams)
		ctx = runtime.NewServerMetadataContext(ctx, md)
		if err != nil {
			runtime.HTTPError(ctx, mux, outboundMarshaler, w, req, err)
			return
		}

		forward_PostgresCrud_List_0(ctx, mux, outboundMarshaler, w, req, resp, mux.GetForwardResponseOptions()...)

	})

	mux.Handle("GET", pattern_PostgresCrud_Read_0, func(w http.ResponseWriter, req *http.Request, pathParams map[string]string) {
		ctx, cancel := context.WithCancel(req.Context())
		defer cancel()
		inboundMarshaler, outboundMarshaler := runtime.MarshalerForRequest(mux, req)
		rctx, err := runtime.AnnotateContext(ctx, mux, req)
		if err != nil {
			runtime.HTTPError(ctx, mux, outboundMarshaler, w, req, err)
			return
		}
		resp, md, err := request_PostgresCrud_Read_0(rctx, inboundMarshaler, client, req, pathParams)
		ctx = runtime.NewServerMetadataContext(ctx, md)
		if err != nil {
			runtime.HTTPError(ctx, mux, outboundMarshaler, w, req, err)
			return
		}

		forward_PostgresCrud_Read_0(ctx, mux, outboundMarshaler, w, req, resp, mux.GetForwardResponseOptions()...)

	})

	return nil
}

var (
	pattern_PostgresCrud_Create_0 = runtime.MustPattern(runtime.NewPattern(1, []int{2, 0, 2, 1}, []string{"v1", "postgrescrud"}, ""))

	pattern_PostgresCrud_List_0 = runtime.MustPattern(runtime.NewPattern(1, []int{2, 0, 2, 1, 2, 2}, []string{"v1", "postgrescrud", "all"}, ""))

	pattern_PostgresCrud_Read_0 = runtime.MustPattern(runtime.NewPattern(1, []int{2, 0, 2, 1, 1, 0, 4, 1, 5, 2}, []string{"v1", "postgrescrud", "id"}, ""))
)

var (
	forward_PostgresCrud_Create_0 = runtime.ForwardResponseMessage

	forward_PostgresCrud_List_0 = runtime.ForwardResponseMessage

	forward_PostgresCrud_Read_0 = runtime.ForwardResponseMessage
)


server.go
package main

import (
	"context"
	"flag"
	"fmt"
	"log"
	"net"

	"github.com/atreya2011/grpc-postgres-crud/postgrescrud"
	"github.com/golang/protobuf/ptypes/empty"
	"github.com/jinzhu/gorm"
	_ "github.com/jinzhu/gorm/dialects/postgres"
	"google.golang.org/grpc"
)

type service struct {
	DB *gorm.DB
}

var grpcAddrFlag = flag.String("addr", ":7000", "Address host:port")

func main() {
	log.Printf("grpc server start on port %v", *grpcAddrFlag)
	// Step 1. listen for connections on tcp
	lis, err := net.Listen("tcp", *grpcAddrFlag)
	// Always handle errors
	if err != nil {
		log.Fatalf("frak")
	}
	// Step 2. Create a new grpc server instance
	srv := grpc.NewServer()
	// Step 3. Initialize the db and store it in service{}
	serv := initDB()
	// Step 4. Register the Service by passing the new server instance
	// and the server struct created above
	// The register function is present in pb.go
	postgrescrud.RegisterPostgresCrudServer(srv, serv)
	// Step 5. Serve the listener created above
	srv.Serve(lis)
}

// Create The following is the implementation of the Create service
// as defined in the proto file. It can be any implemention.
func (s *service) Create(ctx context.Context, req *postgrescrud.CreatePersonRequest) (*postgrescrud.CreatePersonResponse, error) {
	// create new record by passing req.Person into gorm.Create()
	log.Printf("%T\n", req.Person.MiddleName)
	// Note the use of tx as the database handle once you are within a transaction
	// transactions are used to rollback changes in case of an error
	// this is to prevent creation of new record in case of an error
	tx := s.DB.Begin()
	if tx.Error != nil {
		return nil, tx.Error
	}
	person := &postgrescrud.Person{
		MiddleName: "",
	}
	foo := tx.Model(&postgrescrud.Person{}).Create(person)
	bar := foo.Error
	log.Println(bar)
	if bar != nil {
		log.Println("Error: ", bar)
		log.Println("Rollback: ", tx.Rollback().Error)
		return &postgrescrud.CreatePersonResponse{Id: 0}, bar
	}

	// return response
	return &postgrescrud.CreatePersonResponse{Id: 1}, tx.Commit().Error
}

// Read The following is the implementation of the Read service
// as defined in the proto file. It can be any implemention.
func (s *service) Read(ctx context.Context, req *postgrescrud.ReadPersonRequest) (*postgrescrud.ReadPersonResponse, error) {
	// find person with id
	// initialize p with a point to person type
	p := new(postgrescrud.Person)
	// store the result in p using gorm.Where
	err := s.DB.Where("id = ?", req.GetId()).First(p).Error
	if err != nil {
		log.Println(err)
	}
	// return response
	return &postgrescrud.ReadPersonResponse{Person: p}, nil
}

// List lists all the fullnames. Request is empty.
// Below is an example implementation
func (s *service) List(ctx context.Context, e *empty.Empty) (*postgrescrud.ListPeopleResponse, error) {
	// initialize slice of pointers to person
	var people []*postgrescrud.Person
	// pass this to gorm.Find() to fill it with results
	err := s.DB.Find(&people).Error
	if err != nil {
		log.Println(err)
	}
	return &postgrescrud.ListPeopleResponse{Peoples: people}, nil
}

// Delete deletes record based on id
// below is an example implementation
// func (*service) Delete() {
// 	// Delete - delete product
// 	db.Delete(&product)
// }

func initDB() *service {
	const (
		host   = "localhost"
		port   = 5432
		user   = "atreya"
		dbname = "atreya"
	)
	// Step 1. Build the parameter string based on const declared above
	psqlInfo := fmt.Sprintf("host=%s port=%d user=%s dbname=%s sslmode=disable", host, port, user, dbname)
	// Step 2. Connect to database with above parameters
	log.Println(psqlInfo)
	db, err := gorm.Open("postgres", psqlInfo)
	if err != nil {
		panic("failed to connect database")
	}

	// Step 3. Migrate the schema
	db.DropTableIfExists(&postgrescrud.Person{})
	db.AutoMigrate(&postgrescrud.Person{})

	// Step 4. Create a new service instance and store the db
	s := &service{db}

	// return the db
	return s
}
